### PR TITLE
Fixes sofia look up bug in LIVD table

### PR DESCRIPTION
**Pull request by MauriceReeves-usds**
This PR fixes an issue with some test results getting incorrectly encoded as influenza when they're actually from a test that can detect not only COVID but also Influenza A & B. An example is the Sofia 2 test. 

## Changes
- Adds additional methods to `LookupTable` that lets us filter the table before then doing the lookup
- Baking in the filtering logic in the LIVD lookup to only compare based on the `test_performed_code`

To understand why this is necessary, when we look up the test type, we look up by the equipment model name, which in the case of Sofia 2 is: `Sofia 2 Flu + SARS Antigen FIA*`

If you search for that in the LIVD table, you will see there are three rows, one for the COVID test, and one each for influenza A and influenza B. The easiest distinguishing feature to filter further is the `test_performed_code`, which is provided by senders. By using a combination of the two fields, we are able to eliminate the chance that we end up returning incorrect results for a different test type.

### Testing
- [x] Tested locally?
- [x] Ran `quickTest all`?
- [x] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [x] Added tests?

### Process
- [ ] Includes a summary of what a code reviewer should verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
*List GitHub issues this PR fixes*
- #1217

### CODE REVIEW

File: prime-router/src/main/kotlin/LookupTable.kt [hunk 1]
```diff
@@ -57,6 +57,21 @@ class LookupTable(
...
+        indexColumn: String,
+        indexValue: String,
+        lookupColumn: String,
+        filters: Map<String, String>,
+        ignoreCase: Boolean = true
+    ): String? {
+        val filteredTable = filter(filters, ignoreCase)
+        return filteredTable.lookupValue(indexColumn, indexValue, lookupColumn, ignoreCase)
+    }
+
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/LookupTable.kt:74]: Adds a method that takes a map of filters and winnows down the LookupTable before then looking up values

File: prime-router/src/main/kotlin/LookupTable.kt [hunk 2]
```diff
@@ -164,6 +194,22 @@ class LookupTable(
...
+                filters.all { (k, v) ->
+                    val filterColumNumber = headerIndex[k.lowercase()] ?: error("$k doesn't exist in lookup table")
+                    row[filterColumNumber].equals(v, ignoreCase)
+                }
+            }
+        val headerAndFilteredRows = listOf(headerRow) + filteredRows
+        return LookupTable(headerAndFilteredRows)
+    }
+
     /**
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/LookupTable.kt:213]: Creates the method that filters the table and then returns a whole new LookupTable object to operate on

File: prime-router/src/main/kotlin/Mappers.kt [hunk 3]
```diff
@@ -225,16 +225,23 @@ class LIVDLookupMapper : Mapper {
...
+        return listOf(EQUIPMENT_MODEL_NAME, DEVICE_ID, EQUIPMENT_MODEL_ID, TEST_KIT_NAME_ID, TEST_PERFORMED_CODE)
     }
 
     override fun apply(element: Element, args: List<String>, values: List<ElementAndValue>): String? {
+        // get the test performed code
+        val testPerformedCode = values.firstOrNull { it.element.name == TEST_PERFORMED_CODE }?.value
+        val filters: MutableMap<String, String> = mutableMapOf()
+        if (!testPerformedCode.isNullOrEmpty()) {
+            filters[LIVD_TEST_PERFORMED_CODE] = testPerformedCode
+        }
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/Mappers.kt:237]: Get the `test_performed_code` and create a map of filter values for this table to work with

File: prime-router/src/main/kotlin/LookupTable.kt (hunk 1 above)
Comment (RickHawesUSDS) [reply prime-router/src/main/kotlin/LookupTable.kt:74]: Suggest that you update the comment to include examples of how to use the filter. A Map<String, String> is a pretty generic type and doesn't communicate what the filter does.

Comment (RickHawesUSDS) [reply prime-router/src/main/kotlin/LookupTable.kt:74]: Would it better to think of this a second index or third index?

**Comment by RickHawesUSDS** [comment]
So if I understand your solution to the problem, it relies on the `test_performed_code` being present to resolve between the three rows. This works for SimpleReport, but it doesn't for STRAC. Any suggestions on how we solve this for strac?

File: prime-router/src/main/kotlin/LookupTable.kt (hunk 1 above)
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/LookupTable.kt:74]: It is potentially an index, but I didn't want to leave it limited to one or two options max. I wanted to add in the flexibility to get in a bunch of filters if we need to in the future.

**Comment by MauriceReeves-usds** [comment]
@RickHawesUSDS I wasn't even aware that STRAC doesn't send the test_performed_code. We don't map that for them? I'll look again. One suggestion that @jimduff-usds and I thought about was actually splitting the LIVD table by topic, so the two Sofia influenza tests would go into a new file, and then only the COVID tests would be in a covid-19 file, and so on. I'm not super enthused by that option, and it introduces some future complexity, but the other option I was thinking about was using a regex and letting us do a negative match on influenza, which I'm not super crazy about either.

File: prime-router/src/main/kotlin/LookupTable.kt [hunk 4]
```diff
@@ -57,6 +57,21 @@ class LookupTable(
         return table[rowNumber][colNumber]
     }
 
+    /**
+     * Filters a table down based on a predicate, and then invokes the standard lookupValue above
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/LookupTable.kt:61]: Love you added kdoc.  Maybe add the arguments and return too?

File: prime-router/src/main/kotlin/LookupTable.kt [hunk 5]
```diff
@@ -164,6 +194,22 @@ class LookupTable(
             .map { row -> row[selectColumnNumber] }
     }
 
+    fun filter(
```
Comment (carlosfelix2) [inline prime-router/src/main/kotlin/LookupTable.kt:197]: Add kdoc to this one too

**Review by carlosfelix2 — APPROVED** [review]

**Comment by RickHawesUSDS** [comment]
So, on the STRAC problem, I could see two options: 

1. Require them to give us a `test_performed_code`, especially if they start supporting Sofia 2. 
2. Build in logic that distinguishes covid tests from other tests

**Comment by MauriceReeves-usds** [comment]
@RickHawesUSDS my preference is to ask them to start providing the code, since that is the most reliable way to distinguish between the different test types. If not, the other options are:
1. We add a column to the LIVD table for topic, and then we filter down by the topic
2. We split the LIVD table by topic, and load the appropriate one for the schema we're working on
3. We add some kind of in-string or regex matching that eliminates anything that matches a certain pattern/only includes what matches the pattern

**Pull request merged**